src/dials/command_line/correlation_matrix.py

```diff
@@ -0,0 +1,143 @@
+    parse_multiple_datasets,
+)
+from dials.util.options import ArgumentParser, reflections_and_experiments_from_files
+from dials.util.version import dials_version
+
+logger = logging.getLogger("dials.algorithms.correlation.analysis")
+
+phil_scope = iotbx.phil.parse(
+    """\
+include scope dials.algorithms.correlation.analysis.phil_scope
+
+seed = 42
+  .type = int(value_min=0)
+
+output {
+  log = dials.correlation_matrix.log
+    .type = str
+  html = dials.correlation_matrix.html
+    .type = path
+  cc_json = dials.correlation_matrix_cc.json
+    .type = str
+  cos_json = dials.correlation_matrix_cos.json
+    .type = str
+}
+""",
+    process_includes=True,
+)
+
+
+help_message = """
+This module will implement a subset of methods used in dials.cosym to perform
+correlation and cos angle based clustering methods. Data should be passed
+through dials.cosym first to implement consistent symmetry.
+
+Examples::
+
+  dials.correlation_matrix symmetrized.expt symmetrized.refl
+
+"""
+
+
+@show_mail_handle_errors()
+def run(args=None):
+    usage = "dials.correlation_matrix [options] symmetrized.expt symmetrized.refl"
+
+    parser = ArgumentParser(
+        usage=usage,
+        phil=phil_scope,
+        read_reflections=True,
+        read_experiments=True,
+        check_format=False,
+        epilog=help_message,
+    )
+
+    params, options, args = parser.parse_args(
+        args=args, show_diff_phil=False, return_unhandled=True
+    )
+
+    # Configure the logging
+    log.config(verbosity=options.verbose, logfile=params.output.log)
+
+    logger.info(dials_version())
+
+    # Log the diff phil
+    diff_phil = parser.diff_phil.as_str()
+    if diff_phil != "":
+        logger.info("The following parameters have been modified:\n")
+        logger.info(diff_phil)
+
+    if params.seed is not None:
+        flex.set_random_seed(params.seed)
+        np.random.seed(params.seed)
+        random.seed(params.seed)
+
+    if not params.input.experiments or not params.input.reflections:
+        parser.print_help()
+        sys.exit()
+
+    reflections, experiments = reflections_and_experiments_from_files(
+        params.input.reflections, params.input.experiments
+    )
+
+    reflections = parse_multiple_datasets(reflections)
+    if len(experiments) != len(reflections):
+        raise Sorry(
+            "Mismatched number of experiments and reflection tables found: %s & %s."
+            % (len(experiments), len(reflections))
+        )
+    if len(experiments) == 1:
+        raise Sorry(
+            "At least 2 datasets are needed for cluster analysis. Please re-run with more datasets."
+        )
+    try:
+        experiments, reflections = assign_unique_identifiers(experiments, reflections)
+        matrices = CorrelationMatrix(
+            experiments=experiments, reflections=reflections, params=params
+        )
+    except ValueError as e:
+        raise Sorry(e)
+
+    matrices.calculate_matrices()
+    matrices.output_json()
```

Comment:
For the outputs - do we want the json output to be optional, and could they be combined into the same file? A commond dials pattern is to have a parameter `output.json=False`, then you provide a filename to this in order to save the json. If the json includes all the data in the html report then this might be even better for further analysis. At the moment, the dataset numbers in the json/html are not directly linked to the actual experiment identifiers such that you could use this to do some selection on an experiment list. Would it be more useful to have a means to output the linkage matrices (with associated dataset experiment identifiers) to enable futher analysis and cluster selection?

Reply:
That would be useful to feed into other software for future analysis. The output json now has information about the cc/cos matrices, as well as the linkage matrices for both clustering outputs (with ids converted to the proper experiment identifiers) - see f484d11